wallet-next/src/sync.rs

```diff
@@ -0,0 +1,93 @@
+use penumbra_chain::CompactBlock;
+use penumbra_crypto::{
+    merkle::{Frontier, Tree},
+    FullViewingKey, Note, Nullifier,
+};
+use penumbra_proto::Protobuf;
+use penumbra_transaction::action::output;
+
+/// Corresponds to a row in the notes table
+pub struct NoteRecord {
+    pub note_commitment: Vec<u8>,
+    pub height_spent: Option<u64>, //None if unspent, otherwise spent at height_spent
+    pub height_created: u64,
+    // note contents themselves:
+    pub diversifier: Vec<u8>,
+    pub amount: u64,
+    pub asset_id: Vec<u8>,
+    pub transmission_key: Vec<u8>,
+    pub blinding_factor: Vec<u8>,
+    // precomputed decryption of the diversifier
+    pub diversifier_index: Vec<u8>,
+    // the nullifier for this note, used to detect when it is spent
+    pub nullifier: Vec<u8>,
+}
```

Comment:
Let's move this out to `src/note_record.rs` and make the `NoteRecord` be a domain type for the [`NoteRecord` proto type](https://rustdoc.penumbra.zone/main/penumbra_proto/wallet/struct.NoteRecord.html) introduced in #732:
```rust
pub struct NoteRecord {
    pub note_commitment: note::Commitment,
    pub note: Note,
    pub diversifier_index: DiversifierIndex,
    pub nullifier: Nullifier,
    pub height_created: u64,
    pub height_spent: Option<u64>,
}

impl Protobuf<pb::NoteRecord> for NoteRecord {}
```

Folding the note contents up into the existing `Note` domain type adds a tiny amount of code when moving in/out of the database, since the `Note` has to [be built `from_parts`](https://rustdoc.penumbra.zone/main/penumbra_crypto/note/struct.Note.html#method.from_parts), but the upshot is saving a lot of code for protoizing the data for downstream consumers.